### ISSUE TITLE
text upload doesn't upload with options

### ISSUE THREAD

**Issue by antoinegrelard** [cloudinary/cloudinary_npm#294]
Hi!

I had an issue while trying to upload custom text styles through the nodejs API (https://cloudinary.com/documentation/upload_images#creating_images_from_text).

I found out what the issue was but still have a question about the behaviour.

So the issue is that, the text function in uploader.js filters the user's options with the utils.only function, but the second argument is not spreaded (should be ...TEXT_PARAMS instead of TEXT_PARAMS, according to the comments above the utils.only function).

My question is the following: Why did you choose to spread ? What's the advantage in spreading the object argument instead of just passing the object and looping on it normally ?

Thanks !

**Comment by leptians** [comment]
Hi Antoine,

I would like to understand more of the use-case and issue that you had.
Can you share a sample code that caused an issue and how did you fix it on your end?

Regards,
Erwin Lukas

**Comment by antoinegrelard** [comment]
Sure.

I was trying to upload a text style like so :

```
cloudinary.uploader.text(
  "Last name Regular Story",
  {
    public_id: "lastname_regular_story",
    font_family: "Source Sans Pro",
    font_weight: "black",
    font_size: 200,
    font_color: "#E0022A",
  },
);
```

The text was uploaded but not with the styling options I passed it.

Checking the function in uploader.js (line 197), this is what I saw :

```
const TEXT_PARAMS = ["public_id", "font_family", "font_size", "font_color", "text_align", "font_weight", "font_style", "background", "opacity", "text_decoration"];

exports.text = function text(text, callback, options = {}) {
  return call_api("text", callback, options, function () {
    let textParams = utils.only(options, TEXT_PARAMS); // This is where the issue is
    let params = {
      timestamp: utils.timestamp(),
      text: text,
      ...options
    };
    return [params];
  });
};
```

Digging deeper, I checked the utils.only function in the index.js file of the utils folder (line 1349) :

```
 * Returns a new object with key values from source based on the keys.
 * `null` or `undefined` values are not copied.
 * @private
 * @param {object} source The object to pick values from.
 * @param {...string} keys One or more keys to copy from source.
 * @return {object} A new object with the required keys and values.
 */
exports.only = function only(source, ...keys) {
  let result = {};
  if(source) {
    for (let j = 0; j < keys.length; j++) {
      let key = keys[j];
      if (source[key] != null) {
        result[key] = source[key];
      }
    }
  }
  return result;
};
```

We can see that this function expect a bunch of keys as the parameters of the function.

If you go back to where it's used in the text upload function, you'll see that the TEXT_PARAMS array is not spreaded to expose all the keys as arguments, which was the issue.

To correct it, I simply spread the array in the function like so :

```
let textParams = utils.only(options, ...TEXT_PARAMS);
```


Hope it helps !

Antoine

**Comment by igy** [comment]
Hi Antoine,

Thanks for the detailed response. The 'text' method is a feature that's used relatively infrequently, so it's quite possible this is an oversight in the SDK which has not been noticed before.

I've had a look and so far I think that what you've said about the problem is accurate and your proposed solution is the correct fix, so thank you for bringing this to our attention and apologies for the inconvenience.

I've scheduled some time to do some additional testing on this soon. if your assessment is correct and there are no unexpected side-effects of the solution you proposed, I"ll make sure the SDK gets updated to account for this in a similar way so that other customers aren't affected in future

Thanks,
Stephen

**Comment by igy** [comment]
HI Antoine,

Sorry for not following up on this earlier; I missed one of the status emails internally and didn't notice this was now complete - I added the changes you suggested in PR #298  and it was released with v1.16.0 of the SDK, released two weeks ago.

Thanks again for bringing this to our attention
-Stephen
